### ISSUE TITLE
Add Task to Remove All Expired Cached Player and Beatmap Info

### ISSUE THREAD

**Issue by ZeroPyrozen** [Meowhal/osu-ahr#46]
## Description
Add automated or manual task to remove all expired cached player and beatmap info to reduce RAM usage. We can add this task to Discord and console command and after it finish running, it gives the log message. I kinda run out of memory wwwww~~~
😄 

![](https://pyrozen.s-ul.eu/Mn7JxpgI)

**Comment by jramseygreen** [comment]
Why don't you just enter things locally to a sqlite database, or even dare I say, a text file

**Comment by ZeroPyrozen** [comment]
> Why don't you just enter things locally to a sqlite database, or even dare I say, a text file

yup I've been thinking about this approach. Also I need to investigate this exception

```
C:\Auto Host Rotate\osu-ahr-master\src\webapi\BeatmapRepository.ts:53
        throw new FetchBeatmapError(FetchBeatmapErrorReason.PlayModeMismatched);
              ^
FetchBeatmapError: PlayModeMismatched
    at BeatmapRepositoryClass.getBeatmap (C:\Auto Host Rotate\osu-ahr-master\src\webapi\BeatmapRepository.ts:53:15)
    at runMicrotasks (<anonymous>)
    at processTicksAndRejections (node:internal/process/task_queues:96:5)
    at async MiscLoader.checkMirror (C:\Auto Host Rotate\osu-ahr-master\src\plugins\MiscLoader.ts:92:15) {
  isFetchBeatmapError: true,
  reason: 2
}
```

**Comment by ZeroPyrozen** [comment]
![](https://cdn.discordapp.com/attachments/897073872732553216/930462444143394876/unknown.png)

I tried running 3 rooms for 1 month straight. Got out of memory hahaha 💀 💀 💀

**Comment by ZeroPyrozen** [comment]
Let me try run it without caching the map and I will report back in 1 month.

**Comment by Meowhal** [comment]
After all, the memory eventually fills up, doesn't it? I'll make it a top priority task when I'm able to write code.

**Comment by ZeroPyrozen** [comment]
> This is the situation I was afraid of. I'll make it a top priority task when I'm able to write code.

Yup, I've tried running it without using map as a cache for a day. This is my current implementation, we will see whether it can stay below 1GB memory in 2 weeks

```ts
    async getBeatmap(mapId: number, mode: PlayMode = PlayMode.Osu, allowConvert: boolean = true): Promise<BeatmapCache> {
        if (this.fetcher == WebApiClient) {
            if (!WebApiClient.available) {
                this.fetcher = this.websiteFetcher;
            }
        }

        if (WebApiClient.available) {
            this.fetcher = this.websiteFetcher;
            let cache = this.tryGetCache(mapId, mode, allowConvert);
            if (cache) return cache;
    
            const set = await this.fetcher.getBeatmapset(mapId);
            if (set.availability.download_disabled == true || set.availability.more_information != null){
                throw new FetchBeatmapError(FetchBeatmapErrorReason.NotAvailable);
            }
            this.cacheMaps(set);
    
            cache = this.tryGetCache(mapId, mode, allowConvert);
            if (cache) return cache;
    
            throw new FetchBeatmapError(FetchBeatmapErrorReason.PlayModeMismatched);
        }
        else{
            const set = await this.fetcher.getBeatmapset(mapId);
            if (set.availability.download_disabled == true || set.availability.more_information != null){
                throw new FetchBeatmapError(FetchBeatmapErrorReason.NotAvailable);
            }
            const now = Date.now();
            for (const map of [...set.beatmaps ?? [], ...set.converts ?? []] as BeatmapCache[]) {
                const key = this.genKey(map.id, map.mode);
                map.fetchedAt = now;
                map.beatmapset = set;
                map.failtimes = { exit: [], fail: [] };
                //Simplify this
                if(map.id == mapId){
                    if(map.mode_int == PlayMode.Osu.id){
                        if(mode == PlayMode.Osu || (mode != PlayMode.Osu && allowConvert)){
                            return map;
                        }
                    }
                    else{
                        return map;
                    }
                }
            }
            throw new FetchBeatmapError(FetchBeatmapErrorReason.PlayModeMismatched);
        }
    }
```

**Comment by ZeroPyrozen** [comment]
I tried my approach that I put above, it's still crashed because run out of memory. I will add more RAM to my virtual machine to avoid this problem.

**Comment by Meowhal** [comment]
I added cache cleaner that discard beatmap caches and history caches.
Old history data was wasting memory. They only needed to be kept for a short period of time, but they were kept permanently. 
Cleaner will discard them every day.

Lobby#playersMap is also permanently retained, but since i have built the system on the assumption that this data will not be destroyed, safely destroying this data is a difficult task.

As an alternative approach, i changed the app to be launched by node instead of ts-node. This change should help to save some memory.

**Comment by ZeroPyrozen** [comment]
This issue has resolved from 6cd9055671fca3812376e14ad50da44a3d733ca4, thank you so much 😄 .
